docs2/site/docs/guides/dataloader.md

```diff
@@ -89,9 +89,34 @@ method of the returned `IDataLoaderResult<T>`. You can use a synchronous or asyn
 `IDataLoaderResult<T>` if you wish to chain dataloaders together. This may result in the field builder's Resolve delegate
 signature looking like `IDataLoaderResult<IDataLoaderResult<T>>`, which is correct and will be handled properly by the execution strategy.
 
+### GetOrAddBatchLoader vs. GetOrAddCollectionBatchLoader
+
+The decision about whether to use the `GetOrAddBatchLoader` or `GetOrAddCollectionBatchLoader` method should be based on whether each key can link to only one or to multiple values:
+
+- use `GetOrAddBatchLoader` for one-to-one relationships, i.e. when each key links to a single value; think for example of resolving the user who made an order: each order maps to one (and only one) user;
+- use `GetOrAddCollectionBatchLoader` for one-to-many or many-to-many relationships, i.e. when each key links to multiple values; think for example of resolving orders for a user: each user may map to (zero or) more orders.
+
+Note that, from a resolver's perspective, one-to-many and many-to-many relationships are exactly the same, since a resolver only sees one side of the relation. Think of orders and products. In theory, a product can appear in multiple orders and an order can contain multiple products, so this is a many-to-many relationship. However, a resolver for `Product.Orders` (or alternatively, `Order.Products`) will only need to care for the one-to-many side of the relation.
+
 ## Examples
 
-This is an example of using a DataLoader to batch requests for loading items by a key. `LoadAsync()` is called by the field resolver for each Order. `IUsersStore.GetUsersByIdAsync()` will be called with the batch of userIds that were requested.
+Below are some example implementations of a DataLoader for different use cases given the following schema:
+
+```graphql
+type Order {
+    User: User! # one-to-one
```

Comment:
```suggestion
    User: User! # many-to-one
```
Again, this is the same relationship as shown below, just in reverse.  Many orders connect to one user.  Many-to-one.

One-to-one relationships exist but are few and far between.  In one of my GraphQL schemas, I have something like this:

```gql
type Product {
    pricingInformation: ProductPricingInfo
}
```

In my schema/database, this is a one-to-one relationship.  Each 'Product' connects to exactly one 'ProductPricingInfo' and vice versa.

Reply:
Fair enough. I think we just need to agree on whether we want to describe this from a logical viewpoint or simply from the  resolver's "perspective". In the former case, this is indeed a many-to-one relationship, in the latter it can be seen as a one-to-one. My wording is due to my understanding that what matters for the correct usage of the data loader pattern is the relationship as seen from the resolver.

Reply:
By reading further about data loaders outside of this lib, it seems to me that a data loader is not bound to the relationships present in the logical model. One may very well register a generic batch-loading method for orders and use it when resolving a single order from a user (by calling the `GetOrAddBatchLoader` method) and resolve the following query via a data loader:

```graphql
query {
    user {
        order(id: 1) { # batched
            amount
        }
        order(id: 2) { # batched
            amount
        }
```

In this case the "true" relationship between user and order is a one-to-many (or many-to-one as seen from the order side), but the call is to the `GetOrAddBatchLoader` as the resolver here is only concerned with a single order (i.e. a one-to-one relationship).

Reply:
I’m on the road and have not read your revised paragraphs yet, so bear with me. 

I think while it makes sense from a “resolver perspective” to say it’s one-to-one, GraphQL is a relational schema and when we are providing examples that clearly are a many-to-one relationship within the GraphQL context (meaning, the opposite type can a field corresponding to the same relationship), then we should use “many-to-one” in our examples. Or at least put it in parentheses like “(typically this is a many-to-one relationship, etc etc)”.

We have a number of new GraphQL developers that have a good understanding of database design, but need help when figuring out how to configure a GraphQL schema. Stating “one-to-one” for a many-to-one relationship would just add to their confusion. The goal is to make it as simple as possible for new developers to leverage their existing knowledge base and get up to speed with GraphQL quickly. 

I’ll review your updated text as soon as I get a chance. Thanks!

Reply:
> In this case the "true" relationship between user and order is a one-to-many (or many-to-one as seen from the order side)

That's true.

> but the call is to the `GetOrAddBatchLoader` as the resolver here is only concerned with a single order (i.e. a one-to-one relationship).

Only as "seen by the resolver".  Not as the general relationship between users and orders.

Perhaps we should drop the emphasis on the "one-to-many" relationship, moving it to more of a comment, so something like this:

> - `GetOrAddBatchLoader` is used when the resolver is expected to return a single record (or null).  For instance, for a `customer` field for an `Order` type.  Typically these are used for many-to-one relationships, but in some circumstances may represent other relationships.  In this example, the primary key of the customer table is fed to the data loader, which retrieves all customers at a single time from the underlying data source.
> - `GetOrAddCollectionBatchLoader` is used when the resolver is expected to return a list of records.  For instance, for a `orders` field on a `Customer` type.  Typically these are used for one-to-many or many-to-many relationships.  In this example, the primary key of the customer table is fed to the data loader, which retrieves all orders for all matching customer ids at a single time from the underlying data source.

That needs work, but just an idea.  Sorta like some of the text you had I think.

I do think that mentioning the relationship is valuable, as for the typical configuration it is very helpful for developers to leverage their knowledge of their existing database design to use when configuring a GraphQL schema.

Reply:
> Fair enough. I think we just need to agree on whether we want to describe this from a logical viewpoint or simply from the resolver's "perspective". In the former case, this is indeed a many-to-one relationship, in the latter it can be seen as a one-to-one. My wording is due to my understanding that what matters for the correct usage of the data loader pattern is the relationship as seen from the resolver.

I think it's going to be very confusing to developers if we describe it "from the resolver's perspective", using terms that people readily associate with database design.  I'm okay if we want to describe it from the resolver's perspective, but not with terms like "one-to-one" or "many-to-one" to describe it.  Basically what my comment above was saying.

Reply:
I see your concern, though I want to mention that to me, that was exactly the confusing part, as I was reasoning in terms of the relation as defined in the DB, but the API required me to reason in terms of the resolver's point of view :D

In any case, I'll update this so we use the "true" relation and then explain what that means from the resolver's perspective.